### PR TITLE
feat(hotkey): capture data part1 - declare internal collector

### PR DESCRIPTION
**Pull request by Smityz**
According to the comments in hotkey_collector.h and https://github.com/apache/incubator-pegasus/pull/616
`hotkey_coarse_data_collector` is to capture and analyze data in the COARSE state
`hotkey_fine_data_collector` is to capture and analyze data in the FINE state

## Config changes
```diff
[pegasus.collector]
+ hotkey_analyse_time_interval_s
```

### CODE REVIEW

File: src/server/hotkey_collector.cpp [hunk 1]
```diff
@@ -28,10 +30,16 @@ void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request
 
 void hotkey_collector::capture_raw_key(const dsn::blob &raw_key, int64_t weight)
 {
-    // TODO: (Tangyanzhao) Add a judgment sentence to check if it is a raw key
+    dsn::blob hash_key, sort_key;
+    pegasus_restore_key(raw_key, hash_key, sort_key);
+    capture_hash_key(hash_key, weight);
+}
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:36]: add a blank line

File: src/server/hotkey_collector.h [hunk 2]
```diff
@@ -69,6 +84,26 @@ class hotkey_collector
     void capture_hash_key(const dsn::blob &hash_key, int64_t weight);
     void handle_rpc(const dsn::replication::detect_hotkey_request &req,
                     /*out*/ dsn::replication::detect_hotkey_response &resp);
+
+private:
+    detect_result _result;
```
Comment (levy5307) [inline src/server/hotkey_collector.h:89]: I didn't find any usage about `detect_result _result` in this pr, maybe you can add it in next pr.

File: src/server/hotkey_collector.h [hunk 3]
```diff
@@ -69,6 +84,26 @@ class hotkey_collector
...
+    std::unique_ptr<internal_collector_base> collector;
+};
+
+class hotkey_coarse_data_collector : public internal_collector_base
+{
+public:
+    void capture_data(const dsn::blob &hash_key, uint64_t size);
+    void analyse_data();
+    virtual bool get_analysis_result(/*out*/ detect_result &result);
```
Comment (levy5307) [inline src/server/hotkey_collector.h:98]: why did you add `virtual` for this func？

File: src/server/pegasus_server_impl.cpp [hunk 4]
```diff
@@ -31,6 +32,11 @@ namespace server {
 
 DEFINE_TASK_CODE(LPC_PEGASUS_SERVER_DELAY, TASK_PRIORITY_COMMON, ::dsn::THREAD_POOL_DEFAULT)
 
+DSN_DEFINE_int32("pegasus.server",
+                 hotkey_analyse_time_interval_s,
+                 10,
+                 "hotkey_analyse_time_interval_s, in second");
```
Comment (acelyc111) [inline src/server/pegasus_server_impl.cpp:38]:
```suggestion
                 "hotkey analyse interval in seconds");
```

File: src/server/hotkey_collector.h [hunk 5]
```diff
@@ -64,16 +64,34 @@ namespace server {
...
+    virtual void capture_data(const dsn::blob &hash_key, uint64_t weight) = 0;
+    virtual void analyse_data() = 0;
+};
+
+// used in hotkey_collector_state::STOPPED and hotkey_collector_state::FINISHED, avoid null pointers
+class hotkey_empty_data_collector : public internal_collector_base
+{
+public:
+    void capture_data(const dsn::blob &hash_key, uint64_t size);
+    void analyse_data();
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:93]:
```suggestion
    void capture_data(const dsn::blob &hash_key, uint64_t size) {}
    void analyse_data() {}
```

File: src/server/hotkey_collector.cpp [hunk 6]
```diff
@@ -28,10 +36,23 @@ void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request
...
+{
+    // TODO: (Tangyanzhao) add a unit test to ensure data integrity
+    _internal_collector->capture_data(hash_key, weight);
+}
+
+void hotkey_collector::analyse_data() { _internal_collector->analyse_data(); }
+
+void hotkey_empty_data_collector::capture_data(const dsn::blob &hash_key, uint64_t size) {}
+
+void hotkey_empty_data_collector::analyse_data() {}
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:55]:
```suggestion
```

File: src/server/hotkey_collector.cpp [hunk 7]
```diff
@@ -28,10 +36,23 @@ void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request
 
 void hotkey_collector::capture_raw_key(const dsn::blob &raw_key, int64_t weight)
 {
-    // TODO: (Tangyanzhao) Add a judgment sentence to check if it is a raw key
+    dsn::blob temp_raw_key, hash_key, sort_key;
+    temp_raw_key = raw_key;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:40]: `pegasus_restore_key ` will not modify the first parameter, i.e. `raw_key`, is `temp_raw_key` needed?

File: src/server/pegasus_server_impl.cpp [hunk 8]
```diff
@@ -1490,6 +1496,16 @@ ::dsn::error_code pegasus_server_impl::start(int argc, char **argv)
         this, _read_hotkey_collector, _write_hotkey_collector);
     _server_write = dsn::make_unique<pegasus_server_write>(this, _verbose_log);
 
+    ::dsn::tasking::enqueue_timer(LPC_ANALYZE_HOTKEY,
```
Comment (hycdong) [inline src/server/pegasus_server_impl.cpp:1499]: Where is the definition of `LPC_ANALYZE_HOTKEY`?

File: src/server/hotkey_collector.cpp [hunk 9]
```diff
@@ -28,10 +36,18 @@ void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request
...
 {
-    // TODO: (Tangyanzhao) Add a judgment sentence to check if it is a raw key
+    dsn::blob hash_key, sort_key;
+    pegasus_restore_key(raw_key, hash_key, sort_key);
+    capture_hash_key(hash_key, weight);
+}
+
+void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weight)
+{
+    // TODO: (Tangyanzhao) add a unit test to ensure data integrity
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:46]: What is the data integrity to be tested? and how about add the unit test in the pull request?

File: src/server/pegasus_server_impl.cpp (hunk 8 above)
Comment (Smityz) [reply src/server/pegasus_server_impl.cpp:1499]: https://github.com/XiaoMi/rdsn/blob/7601254a6ed140826276da3d35175530fea994f6/include/dsn/dist/replication/replication.codes.h#L74

File: src/server/hotkey_collector.cpp (hunk 9 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:46]: Ok, I will add a function test to test it in the future PR

**Review by hycdong — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]
